FirestoreDocument.js

```diff
@@ -66,40 +66,79 @@ function wrapValue_ (value) {
 
 function unwrapValue_ (value) {
   var type = Object.keys(value)[0]
+  value = value[type]
   switch (type) {
+    case 'referenceValue':
+      // TODO: Allow possibility to retrieve document from this reference
+      // return value.match(/\/documents\/(.*)/)[1];
+      // fall through
+    case 'bytesValue':
     case 'stringValue':
     case 'booleanValue':
     case 'integerValue':
     case 'doubleValue':
-      return value[type]
-    case 'nullValue':
-      return null
+      return value
+    case 'geoPointValue':
+      value = createFirestoreDocument_(value)
+      // Transform coordinates as mapValue object type
+      // fall through
     case 'mapValue':
-      return getFieldsFromFirestoreDocument_(value[type])
+      return getFieldsFromFirestoreDocument_(value)
     case 'arrayValue':
-      return unwrapArray_(value[type]['values'])
-    default:
-      // error
+      return unwrapArray_(value['values'])
+    case 'timestampValue':
+      return new Date(value)
+    case 'nullValue':
+    default: // error
       return null
   }
 }
 
 function wrapString_ (string) {
+  if (regexPath_.test(string)) {
+    return wrapRef_(string)
+  }
+
+  // Not guaranteed to catch all
```

Comment:
Like before, I'd like another comment here.

Also, could you say more about why it's not guaranteed to catch all? Is this something we should be worried about? I'm just doing a read-through so haven't tried testing it out yet, but would like to hear what cases fail here.

If possible, I'd rather not ship something in the library that doesn't always work.